traffic_ops/traffic_ops_golang/server/servers_server_capability.go

```diff
@@ -117,6 +120,22 @@ func (ssc *TOServerServerCapability) Read() ([]interface{}, error, error, int) {
 }
 
 func (ssc *TOServerServerCapability) Delete() (error, error, int) {
+	// Ensure that the user is not removing a server capability from the server
+	// that is required by the delivery services the server is assigned to (if applicable)
+	dsIDs := []int64{}
+	if err := ssc.APIInfo().Tx.QueryRow(checkDSReqCapQuery(), ssc.ServerID, ssc.ServerCapability).Scan(pq.Array(&dsIDs)); err != nil {
+		return nil, fmt.Errorf("checking removing server server capability would still suffice delivery service requried capabilites: %v", err), http.StatusInternalServerError
+	}
+
+	if len(dsIDs) > 0 {
+		dsIdsStr, err := json.Marshal(dsIDs)
+		if err != nil {
+			return nil, fmt.Errorf("formatting response message on bad request to disassociate server capability from server: %v", err), http.StatusInternalServerError
+		}
+		return fmt.Errorf("cannot remove the capability %v from the server %v as the server is assigned to the delivery services %v that require it", *ssc.ServerCapability, *ssc.ServerID, string(dsIdsStr)), nil, http.StatusBadRequest
```

Comment:
This check isn't respecting tenancy; if a user doesn't have tenant permissions over a DS they're not supposed to know it exists. In that case, you should probably give back an Internal Server Error with no user error, I think, to maintain obscurity.

Reply:
based on the result of https://github.com/apache/trafficcontrol/issues/4043 I am unsure if we would want to change this to a 500 or leave it as is

Reply:
This is a user error. 5xx codes are for server errors, it's a violation of HTTP to send them for a client error.

Like I commented in #4043, I think this should be a `403 Forbidden` since we can't really hide DS existence. But if we feel the need to try to obfuscate, a 400 or 404 would be acceptable; but 5xx isn't.

(404 is kind of a little lie; but we can kind of rationalize it as "it doesn't exist, as far as you're concerned." Which IMO is a much smaller lie/violation than a 5xx for a request error. But I'd vote for 400 over 404.)

Reply:
So `api.GenericDelete` probably handles returning 404's for keys that aren't found, huh? I only suggested a 500 because otherwise it seemed like you'd get a 404 **only** when you don't have permission, which isn't an obfuscation at all. Definitely should be a 404 IMO as long as legitimate 404s are possible.

Reply:
My first thought it to leave this as a 400 regardless of that decision ( which 40x code to use).

I am just checking that you can not remove a server capability from a server that is required from an associated ds.  Server server capabilities are not under tenancy so returning a 403 seems a little weird to me.

Reply:
Yeah, 400 is fine.

Reply:
I dont think this code needs to change then - as at the point of getting the affected DSes ids I  am going to return 400 regardless if they have tenancy or not.

One thing I could do is omit the ids from the alert message (either all or the ones they dont have permissions to)

Reply:
I think the ids oughtta be excluded.

Reply:
@ocket8888  I omit the ds IDs the user is not authorized to see from the error message. Can you please rereview the PR?